### ISSUE TITLE
help CMD does not display CMD help section in browser

### ISSUE THREAD

**Issue by kopischke** [fish-shell/fish-shell#3076]
`help CMD` (where CMD is a Fish builtin) will open the default browser to the command documentation index instead of the section for CMD.

### Reproduction Steps:

1. `help function`

### Expected behavior:

The [Fish help `function` section](http://fishshell.com/docs/current/commands.html#function) is displayed in the browser.

### Observed behavior:

The [Fish command help index](http://fishshell.com/docs/current/commands.html) is displayed in the browser.

### Additional information:

- Safari Version 9.1 (11601.5.17.1) is the standard browser
- `$BROWSER` is unset  

---
- Fish version: 2.3, Homebrew installed via `brew install fish --devel`
- Operating system: OS X 10.11.4 _(EDIT: and 10.11.5)_
- Terminal or terminal emulator: Terminal.app

**Comment by faho** [comment]
This works for me on linux, so it's probably an issue with the OSX code path - `help` is a function that does some interesting stuff to find a browser, and on OSX it falls back to `osascript`.

Can you try executing `osascript -e open location "file:///usr/share/doc/fish/commands.html#function"`? That should be what the function ends up executing on your setup.

**Comment by kopischke** [comment]
That command ends giving me the index, i.e. the URL without the anchor. Interestingly, using

```sh
open 'file:///usr/share/doc/fish/commands.html#function'
```

(which would be the most straightforward approach) has the same behavior: it suppresses the anchor.

**Comment by faho** [comment]
In that case, do you have any idea of how we should make this work on OSX? If the tools strip it, there doesn't seem much we can do unless we can get a tool that doesn't.

**Comment by faho** [comment]
A comment in `help`s code says "OS X /usr/bin/open swallows fragments (anchors), so use osascript" - i.e. it was meant to be a workaround for open being borked, and it seems to not work.

**Comment by kopischke** [comment]
Actually, `open` does not swallow anchors in `http://` URLs, at least not on OS X 10.11.4 – but it does so in `file:///` ones. Apparently, this has to do with the internal conversion to a system File URL and [has been present as far back as 2002](http://lists.apple.com/archives/Carbon-development/2002/Jul/msg00071.html).

**Comment by kopischke** [comment]
… and trying to tell OSAscript this is a URL (via `as URL`) doesn’t change a iota of that behavior. Gaaah.

**Comment by kopischke** [comment]
… aaaaaand the change extending this behavior to Standard Additions’ `open location` is obviously recent, as Fish current approach was [still considered state of the art a short while ago](http://apple.stackexchange.com/questions/194501/open-a-file-url-with-anchor-from-terminal).

**Comment by kopischke** [comment]
@faho Looks like calling Standard Additions via a browser host application does work:

```sh 
osascript -e 'tell application "Safari" to open location "file://<path>/file.html#fragment"'
osascript -e 'tell application "Google Chrome" to open location "file://<path>/file.html#fragment"'
# Firefox omitted (not installed) 
```

will open the correct section in Safari and Chrome, respectively. Note this makes no usage of browser specific scripting terminology, it just scopes system `open location` usage. The interesting question is how to handle this situation to provide sane default behavior on OS X…

**Comment by faho** [comment]
@kopischke: If there's no browser-agnostic way to do that it's not worth it, as you could just as easily set $BROWSER and we'd just run that.

**Comment by kopischke** [comment]
> you could just as easily set $BROWSER and we'd just run that

@faho I wish that was true, but on OS X, it’s not. The only way to set `$BROWSER` to a GUI browser is to set it to `open -a "<Browser name>"', and that will ignore anchors, as we have seen. The workaround I showed above is not applicable to `$BROWSER`, as the URL needs to be inserted **inside** the statement, not appended to it ([which is what the `help` function does](https://github.com/fish-shell/fish-shell/blob/master/share/functions/help.fish#L152)).

**Comment by faho** [comment]
So if you just call `chrome https://www.google.com` from fish, that doesn't work? Wow.

**Comment by kopischke** [comment]
> So if you just call chrome https://www.google.com from fish, that doesn't work? Wow.

No you can’t; there isn’t even a `chrome` executable! What there is is a Mach executable located in `/Applications/Google Chrome/Contents/MacOS` that expects to get its arguments via OS X’ LaunchServices. `open` is the  CLI tool to bridge that, which is why it is so maddening it breaks Fish help.

**Comment by jakwings** [comment]
I have no such an issue with `help` or `osascript -e 'open location "<file url>"'` on osx 10.11.0. This may be a bug of osx 10.11.4. Well, osx often produce mysterious bugs after a direct upgrade.

Tested with:
Terminal.app 2.6 (361)
iTerm2 2.9.20160523
\+
Safari 9.0 (11601.1.56)
Chrome 52.0.2743.6 dev
Firefox 46.01

**Comment by krader1961** [comment]
The `osascript -e 'open location "<file url with #anchor>"'` solution works for me on OS X 10.11.6 (latest beta build).

**Comment by kopischke** [comment]
@krader1961 I see, good to know. I will update to OS X 10.11.5 and check if that works again.

@faho the browser agnostic way to do this would be

```fish
set browser (begin; set -lx VERSIONER_PERL_PREFER_32_BIT true; perl -MMac::InternetConfig -le 'print +(GetICHelper "http")[1]'; end)
osascript -e 'tell application "'$browser'" to open location "<file url with #anchor>"'
```

BTW. Let’s hope @krader1961 is right and this monstrosity does not become a permanent addition to `help`.

**Comment by kopischke** [comment]
I have updated my OS and can confirm the issue is still present in OS X 10.11.5. 😞.

**Comment by krader1961** [comment]
When you say the issue is still present are you saying that

```
osascript -e 'open location "file:///usr/local/share/doc/fish/commands.html#function"'
```

doesn't have the expected result or just that `help function` doesn't work?

**Comment by kopischke** [comment]
@krader1961 I mean by that that the issue persists [exactly as I reported it](https://github.com/fish-shell/fish-shell/issues/3076#issue-156980474) after updating to OS X 10.11.5 form OS X 10.11.4.

**Comment by lfiolhais** [comment]
`help function` opens the correct section in the browser for me using 10.11.5

**Comment by krader1961** [comment]
I can't reproduce. Running `help function` and the osascript command directly both work for me. So either the fact I have a lot of homebrew packages installed is a factor or there is something odd about your system, @kopischke. The question is how to determine what that critical difference is.

**Comment by Darkshadow2** [comment]
It works for me on 10.11.4 (haven't updated yet) with both ```help function``` and ```osascript``` and I don't have homebrew installed at all.

@kopischke What is your default browser set to? And have you tried logging in as a different user to see if it still doesn't work there?

**Comment by krader1961** [comment]
@Darkshadow2, He's using Safari according to his original problem statement. My earlier tests were with Chrome. I just tested with Safari as my default browser and it works fine. What I find interesting is @kopischke's [earlier comment](https://github.com/fish-shell/fish-shell/issues/3076#issuecomment-221893692). If I've read this thread correctly the problem on his system is that the anchor fragment in the URL is only ignored if the default browser is used rather than explicitly named. Which is weird.

**Comment by kopischke** [comment]
Argh, turns out it *is* a problem with my configuration: I am using [Choosy](https://www.choosyosx.com) to redirect links for certain services to dedicated SSBs, and that seems to break anchor fragment handing when doing an unscoped `open location` (curiously enough, `tell application "Choosy" to open location` works just fine with anchors). In my defense, I have to say that I thought I had deactivated it in favor of Safari while testing, but obviously, I must have had a brainfart when I did that…

For kicks, I downloaded [Browser Fairy](http://www.browserfairy.com), a Choosy alternative working along the same principles, and it suffers from the same issue: apparently, there is a kink in the Cocoa APIs that kills anchor fragments along the way if you dispatch `file://`URLs. 

Anyway: apologies for the noise, and thanks for all the feedback!
